### PR TITLE
file based user auth

### PR DESCRIPTION
**Pull request by martinlindhe**
DEV-1638

### CODE REVIEW

File: server/api.go [hunk 1]
```diff
@@ -218,6 +222,20 @@ func (al *APIListener) handleGetSessions(w http.ResponseWriter, req *http.Reques
 	al.writeJSONResponse(w, http.StatusOK, response)
 }
 
+// fetching /me returns the currently logged in user and the groups the user belongs to.
+func (al *APIListener) handleGetMe(w http.ResponseWriter, req *http.Request) {
```
Comment (m-terel) [inline server/api.go:226]: Please add API doc for this route in `api-doc.yaml` .

File: server/server.go [hunk 2]
```diff
@@ -38,16 +41,36 @@ func (c *Config) InitRequestLogOptions() *requestlog.Options {
 	return &o
 }
 
+func (c *Config) ReloadAPIAuthBlock() error {
```
Comment (m-terel) [inline server/server.go:44]: nit: please add doc

File: server/server.go [hunk 3]
```diff
@@ -38,16 +41,36 @@ func (c *Config) InitRequestLogOptions() *requestlog.Options {
...
+	block := FileAuth{}
+	if c.APIAuthFilename != "" {
+		b, err := readFileAuth(c.APIAuthFilename)
+		if err != nil {
+			return fmt.Errorf("error parsing %s: %v", c.APIAuthFilename, err)
+		}
+		block = *b
+	}
+
+	user, password, err := parseHTTPAuthStr(c.APIAuth)
```
Comment (m-terel) [inline server/server.go:54]: Please handle the error case. At least log it.

File: server/api_listener.go [hunk 4]
```diff
@@ -146,21 +139,28 @@ func (al *APIListener) handleBearerToken(bearerToken string) (bool, error) {
...
 	}
-
-	user, pass := chshare.ParseAuth(basicAuth)
-	if user == "" || pass == "" {
-		return "", "", fmt.Errorf("can't parse basic-auth string")
+	for _, u := range al.auth.Users {
+		if subtle.ConstantTimeCompare([]byte(basicUser), []byte(u.Name)) == 1 {
+			// plain text password
+			if subtle.ConstantTimeCompare([]byte(basicPwd), []byte(u.Password)) == 1 {
+				al.authenticatedUser = basicUser
```
[the remark this answers is not in the document]
Comment (martinlindhe) [reply server/api_listener.go:151]: Agree. I'll rework this.

File: server/api.go [hunk 5]
```diff
@@ -218,6 +222,20 @@ func (al *APIListener) handleGetSessions(w http.ResponseWriter, req *http.Reques
...
+	me := struct {
+		User   string   `json:"user"`
+		Groups []string `json:"groups"`
+	}{
+		al.authenticatedUser,
```
[the remark this answers is not in the document]
Comment (martinlindhe) [reply server/api.go:231]: Good catch, it needs to be reworked.

File: server/api.go [hunk 6]
```diff
@@ -218,6 +222,20 @@ func (al *APIListener) handleGetSessions(w http.ResponseWriter, req *http.Reques
 	al.writeJSONResponse(w, http.StatusOK, response)
 }
 
+// fetching /me returns the currently logged in user and the groups the user belongs to.
+func (al *APIListener) handleGetMe(w http.ResponseWriter, req *http.Request) {
+	me := struct {
+		User   string   `json:"user"`
```
[the remark this answers is not in the document]
Comment (martinlindhe) [reply server/api.go:228]: From the ticket DEV-1638:

> Fetching /me returns the currently logged in user and the groups the user belongs to.

File: server/api_listener.go [hunk 7]
```diff
@@ -17,8 +17,8 @@ type APIListener struct {
 
 	fingerprint       string
 	connectURL        string
-	authUser          string
-	authPassword      string
+	auth              *FileAuth
```
[the remark this answers is not in the document]
Comment (martinlindhe) [reply server/api_listener.go:20]: Interfaces can be added when needed. We do not know enough of future spec to design them yet, so I rather do a simple implementation for now.

I agree the struct naming is odd, changing it.

File: server/api_listener.go [hunk 8]
```diff
@@ -146,21 +139,28 @@ func (al *APIListener) handleBearerToken(bearerToken string) (bool, error) {
...
-		return "", "", nil
+	if al.auth == nil || len(al.auth.Users) == 0 {
+		return true
 	}
-
-	user, pass := chshare.ParseAuth(basicAuth)
-	if user == "" || pass == "" {
-		return "", "", fmt.Errorf("can't parse basic-auth string")
+	for _, u := range al.auth.Users {
+		if subtle.ConstantTimeCompare([]byte(basicUser), []byte(u.Name)) == 1 {
```
[the remark this answers is not in the document]
Comment (martinlindhe) [reply server/api_listener.go:148]: Refactored bcrypt compare and added a test.

[the remark this answers is not in the document]
Comment (martinlindhe) [reply server/api_listener.go:148]: The function also compares non-encrypted passwords so I don't agree with that naming.
I did refactor the bcrypt compare and added a test.

File: server/api_listener.go [hunk 9]
```diff
@@ -146,21 +139,28 @@ func (al *APIListener) handleBearerToken(bearerToken string) (bool, error) {
...
+			// plain text password
+			if subtle.ConstantTimeCompare([]byte(basicPwd), []byte(u.Password)) == 1 {
+				al.authenticatedUser = basicUser
+				return true
+			}
+
+			// bcrypt
+			if u.Password[0:4] == "$2y$" {
```
[the remark this answers is not in the document]
Comment (martinlindhe) [reply server/api_listener.go:156]: Code now lives in bcryptHashMatches.

File: server/auth.go [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+type FileAuthGroup struct {
+	ID   string
+	Name string
+}
+
+type FileAuthUser struct {
+	ID       string
+	Name     string
+	Password string
+	Groups   []string
```
[the remark this answers is not in the document]
Comment (martinlindhe) [reply server/auth.go:25]: It can be anything. Unique identifiers in text.

File: server/server.go (hunk 2 above)
Comment (martinlindhe) [reply server/server.go:44]: Renamed the function due to struct rename.
 
But comments like this adds 0 value.
```go
// ReloadAPIAuthRepository reloads the APIAuthRepository
func (c *Config) ReloadAPIAuthRepository() error {
    ...
}

File: server/server.go (hunk 3 above)
Comment (martinlindhe) [reply server/server.go:54]: Good catch

Comment (martinlindhe) [reply server/server.go:54]: Good catch

File: server/api.go (hunk 1 above)
Comment (martinlindhe) [reply server/api.go:226]: Done

Comment (martinlindhe) [reply server/api.go:226]: Done

File: server/server_test.go [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+import (
+	"encoding/base64"
+	"encoding/json"
+	"log"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/cloudradar-monitoring/rport/server/ports"
+	"gotest.tools/assert"
```
[the remark this answers is not in the document]
Comment (martinlindhe) [reply server/server_test.go:12]: It was auto imported from using assert. Sure, I'll switch.

File: server/api.go [hunk 12]
```diff
@@ -312,3 +330,11 @@ func (al *APIListener) handleDeleteSessionTunnel(w http.ResponseWriter, req *htt
 	response := successAPIResponse{Success: 1}
 	al.writeJSONResponse(w, http.StatusOK, response)
 }
+
+func GenerateJWTSecret() string {
+	data := make([]byte, 10)
+	if _, err := rand.Read(data); err != nil {
+		log.Fatalf("can't generate API JWT secret: %s", err)
```
[the remark this answers is not in the document]
Comment (martinlindhe) [reply server/api.go]: This code is not new to this PR, but done.

File: server/api_listener.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (martinlindhe) [reply server/api_listener.go:147]: This code is written by @nikita-vanyasin, maybe he can answer that.

File: server/auth.go [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+	user, pass := chshare.ParseAuth(basicAuth)
+	if user == "" || pass == "" {
+		return "", "", fmt.Errorf("can't parse basic-auth string")
+	}
+
+	return user, pass, nil
+}
+
+func bcryptHashMatches(hashedPassword, password string) bool {
+	if len(hashedPassword) > 4 && hashedPassword[0:4] == "$2y$" {
```
[the remark this answers is not in the document]
Comment (martinlindhe) [reply server/auth.go]: From the ticket:
> The password is expected to be bcrypt hashed and it must be compatible with the passwords generated by
> `htpasswd -bnBC 10 "" password | tr -d ':'`

$2y$ is documented here: https://httpd.apache.org/docs/2.4/misc/password_encryptions.html
Added a comment to the function.

File: server/ports/port_distributor.go [hunk 14]
```diff
@@ -62,3 +65,40 @@ func listBusyPorts() (mapset.Set, error) {
...
+	result := mapset.NewThreadUnsafeSet()
+	values := strings.Split(val, ",")
+	for _, v := range values {
+		rangeParts := strings.Split(v, "-")
+		if len(rangeParts) == 1 {
+			result.Add(tryParsePortNumber(rangeParts[0]))
+		} else if len(rangeParts) == 2 {
+			result = result.Union(tryParsePortNumberRange(rangeParts[0], rangeParts[1]))
+		} else {
+			log.Fatalf("can't parse exclude-ports parameter: incorrect range %s", v)
```
[the remark this answers is not in the document]
Comment (martinlindhe) [reply server/ports/port_distributor.go:79]: Again this is not new code from this PR. It has simply been moved and exported.

File: server/server.go [hunk 15]
```diff
@@ -38,16 +41,40 @@ func (c *Config) InitRequestLogOptions() *requestlog.Options {
 	return &o
 }
 
+// ReloadAPIAuthRepository reloads the APIAuthRepository
+func (c *Config) ReloadAPIAuthRepository() error {
+	block := APIAuthRepository{}
```
[the remark this answers is not in the document]
Comment (martinlindhe) [reply server/server.go]: Done

File: server/ports/port_distributor.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (martinlindhe) [reply server/ports/port_distributor.go]: Done

**Comment by nikita-vanyasin** [comment]
@martinlindhe  hey
PR #29 brings changes similar to your regarding 'excluded ports' parsing implementation. It's moved to 'ports' package and doesn't call `.Fatal` functions, returning errors instead. Probably you can wait till #29 merged to master and then rebase

**Comment by m-terel** [comment]
Reimplemented in #37 because the current implementation doesn't work in case of simulates requests , doesn't cover bearer token auth and some other issues.

**Pull request closed without merging**